feat 267: route.long_name should not contain route.short_name

briandonahue opened the pull request:
**Summary:**

Closes #267. Based on implementation suggestion by @aababilov [here](https://github.com/MobilityData/gtfs-validator/pull/501#discussion_r535506016)

> check that long name does not start from the short name followed by ' ', '-' or '('.

**Expected behavior:** 

Should generate WARNING level notice that `route.long_name` should not contain `route.short_name` as referenced in Best Practices: https://gtfs.org/schedule/best-practices/#routestxt

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s) (N/A)

bdferris-v2 commented on main/src/main/java/org/mobilitydata/gtfsvalidator/validator/RouteNameValidator.java:
```diff
@@ -25,13 +27,23 @@
+ *   <li>&lt;a href="https://gtfs.org/schedule/best-practices/#routestxt"&gt;GTFS routes.txt Best
+ *       Practices&lt;/a&gt;
+ * </ul>
+ *
```
This comment block is probably overkill.  If you are looking at `gtfs-validator` source code, you probably know about these resources :)

bdferris-v2 commented on main/src/main/java/org/mobilitydata/gtfsvalidator/validator/RouteNameValidator.java:
```diff
@@ -62,6 +74,22 @@ public void validate(GtfsRoute entity, NoticeContainer noticeContainer) {
+    if (hasLongName && hasShortName) {
+      String longName = entity.routeLongName();
+      String shortName = entity.routeShortName();
+      Pattern pattern = Pattern.compile("^" + shortName + "[\\s\\-\\(]");
```
I think it's a little risky interpreting the route short name as an unescaped regex.  Instead, I do a simple `longName.startsWith(shortName)` check and if that's true, remove the short name prefix from the long name and apply the regex to the remainder.

bdferris-v2 commented on main/src/main/java/org/mobilitydata/gtfsvalidator/validator/RouteNameValidator.java:
```diff
@@ -62,6 +74,22 @@ public void validate(GtfsRoute entity, NoticeContainer noticeContainer) {
+      String longName = entity.routeLongName();
+      String shortName = entity.routeShortName();
+      Pattern pattern = Pattern.compile("^" + shortName + "[\\s\\-\\(]");
+      Matcher matcher = pattern.matcher(longName);
```
I don't think you are using any specific features of Matcher (e.g. capture groups), so I'd just use String.matches() directly.

isabelle-dr reviewed (changes requested):
Thank you for working on this @briandonahue !
Your new validator notice replaces https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md#route_short_and_long_name_equal. Can you remove it in this PR, so that we can close #267? 

We also need the update in RULES.md :)

briandonahue commented:
@isabelle-dr changing this to a draft to avoid accidental merging - realizing we could use a few more test cases and document good/bad data examples

briandonahue commented:
@bdferris-v2 Any thoughts on using parameterized tests where they make sense?  For this where there are many ways the `short_name` can be contained in the `long_name`, it seems like a perfect candidate. Plus some way off day in the future you might be able to use tests params as sample good/bad data in the generated documentation :) 

I see [this library](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-params) which looks useful, but wondering if you had any other recommendations or concerns with doing this.

briandonahue commented:
@isabelle-dr @bdferris-v2 I've made some significant changes. Documenting the rule made me realize it was missing some use cases and that it may not be perfectly clear and/or covering all potential issues.

So far, I've updated to account for short/long are equal (case insensitive), long cannot start with short and be followed by a space, or disallowed character `(`, `-`. I added `)` and am wondering if `:` or other characters should be added?

Here are my current test cases for bad data:
```
           // short, long
            {"L1", "L1 Long Name"},
            {"L1", "L1-Long Name"},
            {"L1", "L1- Long Name"},
            {"L1", "L1 - Long Name"},
            {"L1", "L1)Long Name"},
            {"L1", "L1) Long Name"},
            {"L1", "L1 ) Long Name"},
            {"L1", "L1(Long Name)"},
            {"L1", "L1 (Long Name)"},
            {"L1", "L1( Long Name)"},
            {"L1", "L1(Long Name"},
            {"L1", "L1 (Long Name"},
            {"L1", "L1( Long Name"},
```

Please let me know if any should be removed or added.

briandonahue commented on RULES.md:
```diff
@@ -2181,17 +2182,53 @@ A route's color and `route_text_color` should be contrasting.
+| route3     	| L1                 	| L1- Long Name      	|
+| route3     	| L1                 	| L1 - Long Name     	|
+| route4     	| L1                 	| L1 (Long Name)     	|
+| route5     	| L1                 	| L1) Long Name     	|
```
Let me know if these examples suffice, and/or are representative of bad and good data...

isabelle-dr replied:
Please see my suggestion above. I think the users will get the point with the 3-row table above, although the validator should flag these examples as bad data.

julianharty commented on RULES.md:
```diff
@@ -2181,17 +2182,53 @@ A route's color and `route_text_color` should be contrasting.
+| route4     	| L1                 	| L1 (Long Name)     	|
+| route5     	| L1                 	| L1) Long Name     	|
+
+Example of accetpable data:
```
typo? did you mean acceptable ?

briandonahue replied:
Fixed, thank you!

julianharty commented on RULES.md:
```diff
@@ -2181,17 +2182,53 @@ A route's color and `route_text_color` should be contrasting.
+### route_long_name_contains_short_name
 
-A single route has the same values for `route_short_name` and `route_long_name`.
+Long name can not contain short name for a single route.
```
Why can't the long name contain the short name for a single route? it might be worth explaining the reasoning behind this requirement.

bdferris-v2 replied:
It's a common pattern for most transit applications to display the combination of route_short_name + route_long_name as a combined label for a route, so if the route_long_name contains the short name, it's duplicative.

briandonahue replied:
Should we add to the description to explain this?

julianharty commented on RULES.md:
```diff
@@ -2181,17 +2182,53 @@ A route's color and `route_text_color` should be contrasting.
-A single route has the same values for `route_short_name` and `route_long_name`.
+Long name can not contain short name for a single route.
 
 Example of bad data:
```
Would it be helpful to add the reason(s) why the following examples are considered 'bad data'?

julianharty reviewed (commented):
I don't know enough of this codebase to approve this PR. The code looks clean, I've added some comments/suggestions. Let me know if I can help further.

bdferris-v2 commented on RULES.md:
```diff
@@ -117,6 +117,7 @@ Each Notice is associated with a severity: `INFO`, `WARNING`, `ERROR`.
 | [`platform_without_parent_station`](#platform_without_parent_station)                         | A platform has no `parent_station` field set.                                                                                                                 |
 | [`route_color_contrast`](#route_color_contrast)                                               | Insufficient route color contrast.                                                                                                                            |
+| [`route_long_name_contains_short_name`](#route_long_name_contains_short_name)                                               | Long name can not contain short name for a single route.                                                                                                                |
 | [`route_short_and_long_name_equal`](#route_short_and_long_name_equal)                         | `route_short_name` and `route_long_name` are equal for a single route.                                                                                        |
```
Need to drop `route_short_and_long_name_equal` entry here as well.

bdferris-v2 commented on RULES.md:
```diff
@@ -2181,17 +2182,53 @@ A route's color and `route_text_color` should be contrasting.
+### route_long_name_contains_short_name
 
-A single route has the same values for `route_short_name` and `route_long_name`.
+Long name can not contain short name for a single route.
```
Since this is a Best Practice, not a strict requirement of the spec, I think this should be a "should not", not a "can not".

> The route long name should not contain the short name for a single route.

bdferris-v2 commented on main/src/test/java/org/mobilitydata/gtfsvalidator/validator/RouteNameValidatorTest.java:
```diff
@@ -117,9 +107,52 @@ public void allNamesProvidedAndDifferentFromRouteDescShouldNotGenerateNotice() {
+              {"L1", "L1(Long Name)"},
+              {"L1", "L1 (Long Name)"},
+              {"L1", "L1( Long Name)"},
+            });
```
Since route short name is the same for examples, I think this might be more concisely represented as:

```
String routeShortName = "L1";
List<String> badRouteLongNames = Arrays.asList("L1 Long Name", ...);

for (String badRouteLongName : badRouteLongNames) {
```

isabelle-dr commented on RULES.md:
```diff
@@ -2181,17 +2181,53 @@ A route's color and `route_text_color` should be contrasting.
-A single route has the same values for `route_short_name` and `route_long_name`.
+Long name should not contain short name for a single route.
 
 Example of bad data:
```
```suggestion
Good examples:
| route_short_name/route_long_name                                                                | Dataset                                                                                                                                                                |
| ----------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| ["N"/"Judah"](https://www.sfmta.com/getting-around/transit/routes-stops/n-judah)                | [Muni San Fransisco](https://storage.googleapis.com/storage/v1/b/mdb-latest/o/us-california-san-francisco-municipal-transportation-agency-sfmta-gtfs-50.zip?alt=media) |
| ["6"/"ML King Jr Blvd"](https://trimet.org/schedules/r006.htm)                                  | [Trimet Portland Streetcar](http://developer.trimet.org/schedule/gtfs.zip)                                                                                             |
| ["55"/"Boulevard Saint Laurent"](https://www.stm.info/en/info/networks/bus/local/line-55-north) | [STM Montreal](https://storage.googleapis.com/storage/v1/b/mdb-latest/o/ca-quebec-societe-de-transport-de-montreal-gtfs-1221.zip?alt=media)                            |
| ["1"/"Rangiora/Cashmere"](https://www.metroinfo.co.nz/timetables/1-rangiora-cashmere/)          | [Metro Christchurch](https://storage.googleapis.com/storage/v1/b/mdb-latest/o/nz-christchurch-christchurch-metro-gtfs-1313.zip?alt=media)                              |

Bad examples:
| route_short_name/route_long_name |
|-------------------------------------|
| "604"/"604"                                         |
| "14"/"Route 14"                                    |
| "2"/"Route 2: Bellows Falls In-Town" |

```

isabelle-dr commented on RULES.md:
```diff
@@ -2181,17 +2181,53 @@ A route's color and `route_text_color` should be contrasting.
+| route4     	| L1                 	| L1 (Long Name)     	|
+| route5     	| L1                 	| L1) Long Name     	|
+
+Example of acceptable data:
```
Again, I think we can remove this table if you agree with my suggestion

isabelle-dr commented on RULES.md:
```diff
@@ -2181,17 +2181,53 @@ A route's color and `route_text_color` should be contrasting.
+| route2     	| BANK               	| Bankford          	|
+
+#### References
+* [routes.txt best practices](https://gtfs.org/schedule/best-practices/#routestxt)
```
Please have the Best Practices with cap B and P (something we don't always do)

isabelle-dr commented on RULES.md:
```diff
@@ -2181,17 +2181,53 @@ A route's color and `route_text_color` should be contrasting.
+
+<a name="RouteShortAndLongNameEqualNotice"/>
+
+### route_short_and_long_name_equal
```
to remove, this will be catched by your new rule

isabelle-dr reviewed (commented):
A few comments in-line, thank you so much for working on this @briandonahue!
A great addition to this validator :)

bdferris-v2 commented on main/src/test/java/org/mobilitydata/gtfsvalidator/validator/RouteNameValidatorTest.java:
```diff
@@ -117,9 +107,51 @@ public void allNamesProvidedAndDifferentFromRouteDescShouldNotGenerateNotice() {
+              "L1(Long Name)",
+              "L1 (Long Name)",
+              "L1( Long Name)",
+            });
```
You don't need the new `String[] { }` call here, as Arrays.asList(...) accepts a var-arg parameter list.

bdferris-v2 commented on RULES.md:
```diff
@@ -2231,29 +2231,39 @@ A route's color and `route_text_color` should be contrasting.
 </details>
 
-<a name="RouteShortAndLongNameEqualNotice"/>
+<a name="RouteLongNameConainsShortName"/>
```
This name needs to end with "Notice" to get the documentation test to pass.

bdferris-v2 reviewed (approved):
Modulo the small changes requested, this LGTM.

briandonahue commented:
@bdferris-v2 Made those fixes, thanks.

isabelle-dr commented on RULES.md:
```diff
@@ -2260,29 +2260,39 @@ A route's color and `route_text_color` should be contrasting.
+|-----------------	|---------------------------------------- |--------	|
+| `routeId`       	| The id of the faulty record.           	| String 	|
+| `csvRowNumber`  	| The row number of the faulty record.   	| Long   	|
+| `routeShortName` 	| The `short_name` of the faulty record.  | String 	|
```
Could you please replace `short_name` with `route_short_name` to be consistent with the spec?

isabelle-dr commented:
One last small change before we merge :)

isabelle-dr commented:
Why aren't we getting the acceptance test report for this PR? 
cc @bdferris-v2

bdferris-v2 commented:
@isabelle-dr because the acceptance test workflow only has permission to post the comment with the report summary if the PR is coming from a branch of the `gtfs-validator` repo itself, as opposed to a fork of the repo, as is the case with this PR (and many coming from external contributors).  This is discussed in more detail [here](https://github.com/MobilityData/gtfs-validator/issues/1232#issuecomment-1211091628).  Unfortunately, I've not seen a good way around this.

davidgamez commented:
> @bdferris-v2 Any thoughts on using parameterized tests where they make sense? For this where there are many ways the `short_name` can be contained in the `long_name`, it seems like a perfect candidate. Plus some way off day in the future you might be able to use tests params as sample good/bad data in the generated documentation :)
> 
> I see [this library](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-params) which looks useful, but wondering if you had any other recommendations or concerns with doing this.

That's a good call. Definitely the **bad list** test case is a great example for parameterizing a test. We can have this *"feature"* for _free_ if we decide to upgrade the JUnit version.  [JUnit 5 Parameterized Tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests)
Let's keep the topic alive and have the conversion out of scope of this PR.

davidgamez reviewed (approved):
LGTM!

briandonahue commented:
> @isabelle-dr because the acceptance test workflow only has permission to post the comment with the report summary if the PR is coming from a branch of the `gtfs-validator` repo itself, as opposed to a fork of the repo, as is the case with this PR (and many coming from external contributors). This is discussed in more detail [here](https://github.com/MobilityData/gtfs-validator/issues/1232#issuecomment-1211091628). Unfortunately, I've not seen a good way around this.

@isabelle-dr @bdferris-v2 This was created before I had access to the main repo.  Should I close this PR and open a new one from a local branch?

bdferris-v2 commented:
Two things.  The acceptance test still ran and you can access the logs in the `Checks` tab to see the result:

✅ Rule acceptance tests passed.
New Errors: 0 out of 1419 datasets (~0%) are invalid due to code change, which is less than the provided threshold of 1%.
Dropped Errors: 0 out of 1419 datasets (~0%) are invalid due to code change, which is less than the provided threshold of 1%.
0 out of 1419 sources (~0 %) are corrupted.
Commit: 4a956ffa6e8a4a5ab006ed999bcbdad0444d8a97
Download the full acceptance test report [here](https://github.com/MobilityData/gtfs-validator/actions/runs/43047391[19](https://github.com/MobilityData/gtfs-validator/actions/runs/4304739119/jobs/7508374310#step:7:20)) (report will disappear after 90 days).
✅ Rule acceptance tests passed.

That said, per #1349, the acceptance test only tracks changes to `ERROR` notices, not warnings.  As this PR only changes a warning notice, it wouldn't show up as a change in the acceptance test.  That's something I'd like to fix.

isabelle-dr commented:
@briandonahue good to merge?

briandonahue commented:
> @briandonahue good to merge?

@isabelle-dr Yes! 😄
